### ISSUE TITLE
[bug] Ingest does not work in LXC container (was: Ingested books are not in the library)

### ISSUE THREAD

**Issue by friki67** [crocodilestick/Calibre-Web-Automated#88]
**Describe the bug**
Ingested books are not loaded in the library, but they are deleted from ingest directory.

**To Reproduce**
1. Create a new library using Calibre Desktop, with or without books
2. Set up Calibre Web Automated, all books present in the web ui
3. Put a new book in the ingest directory
4. Book is deleted but not present in the library (not imported) Log in ui-admin-logs shows nothing for this.

**Expected behavior**
Book apperars in the library

**Configuration(please complete the following information):**
 - OS: Incus Ubuntu 24.04 container + docker inside 
 - Hardware: x86 server
 - Version latest docker container

Calibre web worked fine. Incus user id is mapped 1000:1000 and PUID and PGID are set to 1000 too in docker. This setting is fine for Calibre Web.

EDIT: upload books using the upload button worked fine

How to see logs of what is happening when ingesting books?

**Comment by MadManMcIvor** [comment]
I was having this issue as well earlier today. And I think I got it fixed for me, but was trying to verify that my change is actually the issue and not some fluke. 

The recommended docker compose setup has it as :  
      - PUID=1000
      - PGID=100 (one hundred instead of one thousand 1000)
I switched the PGID to 1000 and it ran as I'd expected. Though, I had previously had it working with that 1000:100 setup so I'm not sure what happened to cause it start messing up. But thought it'd be something to double check.

As for the logs, I've seen I have to check two places. There's the container logs which I check in the portainer (where I'm running these containers) logs, and then there's the logs which you can find in the admin panel while running the app. It's near the bottom in the "Administration" section

**Comment by friki67** [comment]
Is is a permissions thing. No matter if I set PGID to 1000 or 100, when launching docker instance I'm getting:
```
[cwa-set-perms] Recursively setting ownership of everything in /config, /calibre-library, /app/calibre-web-automated to abc:abc...
[cwa-set-perms] Successfully set permissions for '/config'!
chown: changing ownership of '/calibre-library/MyAuthor/MyBook (1)/MyBook - MyAuthor.epub': Operation not permitted
chown: changing ownership of '/calibre-library/MyAuthor/MyBook (1)cover.jpg': Operation not permitted
chown: changing ownership of '/calibre-library/MyAuthor/MyBook (1)': Operation not permitted
chown: changing ownership of '/calibre-library/MyAuthor': Operation not permitted
chown: changing ownership of '/calibre-library/.calnotes/retired': Operation not permitted
chown: changing ownership of '/calibre-library/.calnotes/notes.db': Operation not permitted
chown: changing ownership of '/calibre-library/.calnotes/resources': Operation not permitted
chown: changing ownership of '/calibre-library/.calnotes/backup': Operation not permitted
chown: changing ownership of '/calibre-library/.calnotes': Operation not permitted
chown: changing ownership of '/calibre-library/metadata.db': Operation not permitted
chown: changing ownership of '/calibre-library': Operation not permitted
[cwa-set-perms] Service could not successfully set permissions for '/calibre-library' (see errors above).
```
Note that the book was uploaded using CWA, but something happens with the permissions. Maybe if user abc was 1000:1000 and not 1000:100 ... The problem is in the translation of id's outside my incus container (as said I had defined 1000 to 1000 in user and group) ....

It should work. Calibre Web "classic" works, but CWA fails.

Regards

**Comment by jmarmstrong1207** [comment]
try doing chown -R on your calibre library manually. Does this fix it?

**Comment by friki67** [comment]
> try doing chown -R on your calibre library manually. Does this fix it?

No, I've tried chown to one and the other, chmod a+w, and nothing seems to help. 

I've been trying another thing and now I have got ingest working. But ingested books cannot be deleted, and uploaded books can.

The ingest part is working as root. See my calibre-library directory  after upload one book and ingest another. The ingested one is property of root:root, so cannot be deleted using the "Book list" mode from  CWA.
```
drwxr-xr-x.  3 abc  users   4096 Sep 18 19:15 'Adam Nevill'
drwxr-xr-x.  3 root root    4096 Sep 18 19:08 'Adolf Fort'
-rw-rw-rw-.  1 abc  users 397312 Sep 18 19:15  metadata.db
```

**Comment by friki67** [comment]
Incus/LXD container, docker inside, all works except for ingest. It deletes the books in the ingest directory but is unable to import them.
Docker container as a OCI container directly in Incus. Same thing.
I've tried mapping ids to my host user (1000-1000:1000-100) and setting the environment variables as recommended, using 1000:1000, etc. I suspect that Ingest creates the directory as root user and after that chowns it, and from inside the LXC container this not seems to work.

**Comment by friki67** [comment]
Ok, I made it work setting my container as privileged - security.privileged=true in Incus- (you know inside root=outside root) but this is contrary to the containers philosophy (for me security and isolation), isn't it?

**Comment by captainju** [comment]
Do you mind sharing you LXC setup script please ?

**Comment by 0x09AF** [comment]
I am curious about this issue because like in a previous [issue](https://github.com/crocodilestick/Calibre-Web-Automated/issues/30) I could not get it to work because of `chown`. I deployed version 2.1.2 and still the same issue - CWA container tries to 'chown' a directory from an unprivileged LXC container and fails to do so because it's ... well, an unprivileged container. I have user and group 1000:1000 mapped inside to outside of the LXC, and still doesn't work as expected. Manually changing permissions would work but only until the container restart/rebuild

**Comment by davidancalagon** [comment]
I'm having similar issues to this using CWA with a NFS volume.  Permissions on the NFS directory are set for the same uid and gid as defined in the env variables for PUID and GID, but ingest fails to complete due to some sort of chown error during ingest.

**Comment by davidancalagon** [comment]
The error, for reference:

2024-12-26T03:19:50.283339301Z [ingest-processor] An error occurred while attempting to recursively set ownership of /calibre-library/ to abc:abc. See the following error:
2024-12-26T03:19:50.283342217Z Command '['chown', '-R', 'abc:abc', '/calibre-library/']' returned non-zero exit status 1.

Why is it attempting to do a recursive chown for abc:abc when the puid/gid is set for 2003:1002?

**Comment by kennyboy55** [comment]
I just setup CWA 3.1.0 for the first time and am running into the same issue with my NFS mapped folder. It somehow worked for a minute, because a part of my books were ingested and copied over to the `/ebooks` directory, which is an NFS mount. However, it now stopped working again and every time an ingestion is attempted, it tries to `chown` every folder in the `/ebooks` folder. It is trying to set it to the user `'abc:abc'` for whatever reason. I have the user and group set, which correspond to the correct users both inside, outside and on the NFS mount. No other linuxserver images have this issue for me.

```
    environment:
      - PUID=1000
      - PGID=1000
```

```
calibre-web-automated    | chown: changing ownership of '/ebooks': Operation not permitted
calibre-web-automated    | [ingest-processor] An error occurred while attempting to recursively set ownership of /ebooks to abc:abc. See the following error:
calibre-web-automated    | Command '['chown', '-R', 'abc:abc', '/ebooks']' returned non-zero exit status 1.
```

I should also note that I am using the Calibre docker mod from linuxserver, as I do not have a separate Calibre library.
`DOCKER_MODS=lscr.io/linuxserver/mods:universal-calibre-v7.16.0`

Version info:

<img width="495" height="293" alt="Image" src="https://github.com/user-attachments/assets/cd47de7a-12a7-4fa4-8c57-59ae569ce2d3" />  


_Later addition_ 
I just want to mention that I am using the new feature of 3.1.0 where my library is split. The epub files are stored on the `/ebooks` NFS mount, but the `metadata.db` is stored on the SSD of the machine where the docker container is running.

**Comment by kennyboy55** [comment]
After looking a bit more into it, I assume it is caused by the way LinuxServer.io manages user privileges in their containers, see [here](https://docs.linuxserver.io/misc/non-root/).

The inotifywait commands are run with the `abc` user, one of which is looking at the ingest folder. So this is probably why the ingest service is also triggered under that user?
<img width="1416" height="195" alt="Image" src="https://github.com/user-attachments/assets/737b2d0e-5d7a-4cdf-8fa2-6f021bfced68" />

Sadly LinuxServer does not support any configuration which uses NFS. I am not sure how to get around this, and if that is a wise thing to do in the first place.

**Comment by 0x09AF** [comment]
I gave up on this and created a container with local storage. Books do not take as much space as other media.  
None of my other containers have this behavior.

**Comment by kennyboy55** [comment]
I updated to the latest version of Calibre-Web-Automated just now, and it might somehow have been fixed. It seems to work just fine now, even though I changed nothing in the configuration. I tried ingesting a few ebooks and they are copied to the NFS share just fine, and added to the Calibre library. Possibly this has been fixed in 3.1.2, since the project structure was changed at this point (from being a mod on top of Calibre-web, to having everything in the Calibre-web-automated repo).

**Comment by kennyboy55** [comment]
Actually, I did also change one thing in the docker-compose file. Calibre-web-automated now uses Calibre v8.8.0, so I changed the docker mod. That might have fixed some things as well?

```
calibre-web-automated:
    image: crocodilestick/calibre-web-automated:latest
    container_name: calibre-web-automated
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Amsterdam
      - HARDCOVER_TOKEN=<redacted>
      - DOCKER_MODS=lscr.io/linuxserver/mods:universal-calibre-v8.8.0
    volumes:
      - ./cwa-config:/config
      - ./ingest:/cwa-book-ingest
      - ./library:/calibre-library
      - /mnt/storage/books/ebooks:/ebooks
    ports:
      - 8083:8083
    restart: unless-stopped
```
